### ISSUE TITLE
WMS: Resolution-Parameter in Request-URL

### ISSUE THREAD

**Issue by ghschier** [openlayers/openlayers#14391]
The current behaviour (method **getRequestUrl_** in **TileWMS/ImageWMS**) is, that the additional parameters for the screen resolution (MAP_RESOLUTION/DPI) are appended to the WMS-URL only, if the device-pixel-ratio (DPR) is not equal to 1. The used screen resolution is hardcoded to 90.

This parameters should not depend on the DPR. There are also screens with high resolution with a DPR of 1. The parameters should be appended always to the url, so far "serverType" is defined. Further, there should be as setter-method for the screen resolution.

**Comment by ahocevar** [comment]
You can easily opt out of that by not configuring the WMS source with a `serverType`.

**Comment by ghschier** [comment]
If `serverType` is `undefined`, the `pixelRatio` will always be set to 1. That is not what I want...

<pre>
getImageInternal(extent, resolution, pixelRatio, projection) {
    ...
    if (pixelRatio != 1 && (!this.hidpi_ || this.serverType_ === undefined)) {
      pixelRatio = 1;
    }
    ...
</pre>

**Comment by ahocevar** [comment]
Can you elaborate on what you mean by "There are also screens with high resolution with a DPR of 1", and how that relates to the rendered map resolution of a WMS? The former is relevant for the resolution of the map's output canvas, the latter just determine which styling rules are applied when rendering the WMS image on the server.

**Comment by ghschier** [comment]
The higher the screen resolution, the smaller fonts and symbols are displayed on the screen. Mapserver (as a example) assumes a screen resolution of 72DPI. On a laptop with 120DPI or higher fonts and symbols therefore become very tiny. This has nothing to do with the device-pixel-ratio, the DPR is 1 in this case.
If you send your screen-resolution to the WMS by appending the mentioned parameters (and you are lucky and the server supports this), fonts and symbols have their intended size on the screen.

**Comment by ahocevar** [comment]
@ghschier Ok, we're talking about the same thing. You can configure that in the `params` of the WMS source.

**Comment by ghschier** [comment]
No, I can't. The `params` will be overwritten, if DPR <> 1. I can avoid this by omitting the `serverType` in the configuration, but then the `pixelRatio` is set to 1, and this not intended, too.
IMO the behaviour should be like this (in pseudo-code):
<pre>
if `serverType` is defined{
     //append always Resolution-Parameter, if serverType is defined
    res = screenResolution * pixelRatio;  //pixelRatio can be 1
    params['...'] = res; 
}
</pre>
At the moment the screenResolution is hardcoded to 90. It would be nice, if one could set the value manually.

**Comment by ahocevar** [comment]
I noticed that the `hidpi` Parameter is currently not handled as expected. With a simple change, this could be fixed:
```diff
diff --git a/src/ol/source/ImageWMS.js b/src/ol/source/ImageWMS.js
index 26a558fd4..b91c6d249 100644
--- a/src/ol/source/ImageWMS.js
+++ b/src/ol/source/ImageWMS.js
@@ -286,7 +286,7 @@ class ImageWMS extends ImageSource {
 
     resolution = this.findNearestResolution(resolution);
 
-    if (pixelRatio != 1 && (!this.hidpi_ || this.serverType_ === undefined)) {
+    if (pixelRatio != 1 && !this.hidpi_ && this.serverType_ === undefined) {
       pixelRatio = 1;
     }
 
@@ -395,7 +395,7 @@ class ImageWMS extends ImageSource {
       params['STYLES'] = '';
     }
 
-    if (pixelRatio != 1) {
+    if (pixelRatio != 1 && this.serverType_) {
       switch (this.serverType_) {
         case 'geoserver':
           const dpi = (90 * pixelRatio + 0.5) | 0;
```
Then you could configure a source like this:
```js
     source: new ImageWMS({
       url: 'https://ahocevar.com/geoserver/wms',
       params: {'LAYERS': 'topp:states', 'FORMAT_OPTIONS': 'dpi:' + 300 * DEVICE_PIXEL_RATIO},
       hidpi: true,
     }),
```
If you agree with that, you could submit a pull request along these lines.

**Comment by ghschier** [comment]
OK, this should do the trick for `ImageWMS`. `TileWMS` should be modified in an analogue way.
